variance/bias tradeoff and Overfitting (#42)
aml4td/website: chapters/overfitting.qmd

```{r}
#| label: overfitting-setup
#| include: false
#| message: false
#| warning: false

source("../R/_common.R")

# ------------------------------------------------------------------------------
# Required packages

library(tidymodels)
library(bonsai)
library(patchwork)
library(rlang)
library(kernlab)
library(splines2)
library(doParallel)

# ------------------------------------------------------------------------------
# Set options

cl <- makePSOCKcluster(parallel::detectCores(logical = TRUE))
registerDoParallel(cl)
tidymodels_prefer()
theme_set(theme_transparent())
set_options()

# ------------------------------------------------------------------------------
# Load other data

load("../RData/bayes_opt_calcs.RData")
```

This chapter describes the most crucial concept in predictive modeling: overfitting. When a model fails, it is almost always due to overfitting of some sort. The problem is that you may only realize that there is an issue once you finish the development phase of modeling and are exposed to completely new data.

To get started, we'll discuss model complexity and how it can be both good and bad. 

## Model Complexity and Overfitting {#sec-complexity}

> Many predictive models have parameters. For example, the linear regression model initially shown in @eq-poly-linear-reg contains slopes and intercepts (the $\beta$ values) estimated during model training using the predictor and outcome data. In other words, it is possible to write the equations to solve to estimate these values. 

Author: @eq-poly-linear-reg  is in PR #37

Two other models were discussed in @sec-whole-game. First, the tree-based model shown in @fig-reg-tree had five splits of the data. This led to a model fit that resulted in six regions of the $x$ space, each with its specific predicted value. We could have continued to split the training set into finer and finer partitions so that the fitted curve more effectively modeled the observed data. As such, the depth of the tree was a parameter that needed to be set before the model fit could be estimated. 

Similarly, the neural network required the number of hidden units to be defined (among other values). We weren't sure what that should be, so a set of candidate values were assessed, and the one with the best results was used for the final model fit (see @fig-neural-network). We've also seen specific tuning parameters in @sec-feature-hashing, @sec-combining-categories, and also for the embedding methods in @sec-embeddings. 

> In each of these cases, there was a **tuning parameter**  (a.k.a., _hyperparameter_) that defined the structure of the model and could not be directly estimated from the data. In each case, increasing the parameter's value made the model's potential structure more elaborate. The effect of this increase in the model's _complexity_ is additional capability to adapt to any pattern seen in the training data. The straight line fit associated with a linear regression model without polynomial terms would have done an abysmal job on such a nonlinear data set. Adding additional complexity allowed it to better conform to the training set data. 

Reviewer: ```suggestion
In each of these cases, there was a **tuning parameter**  (a.k.a., _hyperparameter_) that defined the structure of the model and could not be directly estimated from the data. Increasing the parameter's value for each model made the model's potential structure more elaborate. The effect of this increase in the model's _complexity_ is additional capability to adapt to any pattern seen in the training data. The straight line fit associated with a linear regression model without polynomial terms would have done an abysmal job on such a nonlinear data set. Adding additional complexity allowed it to better conform to the training set data. 
```

> The problem is that added complexity can also add risk. Being more capable of adapting to subtle trends in the training data is only a good thing when such trends _generalize_ to other data. Recall in @sec-effect-encodings where we were estimating the effect of an agent on the price of a hotel room. One agent had a sinlge booking. The naive estimate of the ADR wouldn't generalize well. As they accrue more and more bookings, their mean would undoubtedly change to a more realistic value. Using the raw ADR value would "overfit to the training set data."

Reviewer: ```suggestion
The problem is that added complexity can also add risk. Being more capable of adapting to subtle trends in the training data is only a good thing when such trends _generalize_ to other data. Recall in @sec-effect-encodings where we were estimating the effect of an agent on the price of a hotel room. One agent had a single booking. The naive estimate of the ADR wouldn't generalize well. As they accrue more and more bookings, their mean would undoubtedly change to a more realistic value. Using the raw ADR value would "overfit to the training set data."
```

For many sophisticated models, the complexity can be modulated via one or more tuning parameters that detect inconsequential patterns in the training data. The problem is that they may go too far and over-optimize the model fit. 

As an example, @fig-two-class-overfit shows a scenario where there are two predictors (A and B) and each point belongs to one of two classes (vermilion circles or blue triangles). These data were simulated and panel (a) shows a thick grey line representing the true class boundary where there is a 50% chance that a data point belongs to either class. Points to the left of the grey curve are more likely to be circles than points to the right. The curve itself has a smooth, parabolic shape. 

> We can fit a model to these data and estimate the class boundary. The model we used^[A support vector machine. See @sec-cls-svm.] has a tuning parameter call the _cost value_. For now, you can think of this parameter as one that discourages the model from making an incorrect prediction (i.e., placing it on the wrong side of the estimated boundary). In other words, as the cost parameter increases, the model is more and more incentivized to be accurate on the training data. 

Reviewer: ```suggestion
We can fit a model to these data and estimate the class boundary. The model we used^[A support vector machine. See @sec-cls-svm.] has a tuning parameter called the _cost value_. For now, you can think of this parameter as one that discourages the model from making an incorrect prediction (i.e., placing it on the wrong side of the estimated boundary). In other words, as the cost parameter increases, the model is more and more incentivized to be accurate on the training data. 
```

  tibble(A = c(2.20),  B = c(-3.5),
         C = c(1.25),  D = c(-3.5))

high_plot <- 
  base_plot  +
  geom_contour(
    data = high_cost,
    aes(z = value),
    breaks = 0,
    col = "black",
    linewidth = 1 / 2,
    alpha = 3 / 4
  )  +
  geom_segment(data = pointers,
               aes(xend = C, yend = D),
               arrow  = arrow(length = unit(0.03, "npc"))) +
  geom_segment(data = pointers_blowback,
               aes(xend = C, yend = D),
               color = "red",
               arrow  = arrow(length = unit(0.03, "npc"))) +
  ggtitle("(d) High complexity")  +
  labs(x = "Predictor A", y = NULL)

mid_plot <- 
  base_plot  +
  geom_contour(
    data = mid_cost,
    aes(z = value),
    breaks = 0,
    col = "black",
    linewidth = 1 / 2,
    alpha = 3 / 4
  )  +
  ggtitle("(c) Medium complexity") +
  labs(x = "Predictor A", y = NULL)

low_plot <- 
  base_plot  +
  geom_contour(
    data = low_cost,
    aes(z = value),
    breaks = 0,
    col = "black",
    linewidth = 1 / 2,
    alpha = 3 / 4
  )  +
  ggtitle("(b) Low complexity") +
  labs(x = "Predictor A", y = NULL)

base_plot + low_plot + mid_plot + high_plot + plot_layout(nrow = 1, byrow = FALSE)
```

> When the cost is increased, @fig-two-class-overfit(c) shows an increase in complexity. The boundary is more parabolic, and a few more training set points are on the correct side of the estimated line. 

Reviewer: ```suggestion
When the cost is increased, the complexity also increases (@fig-two-class-overfit(c)). The boundary is more parabolic, and a few more training set points are on the correct side of the estimated line. 
```

@fig-two-class-overfit(d) shows what occurs when the model is instructed that it is incredibly bad when incorrectly predicting a data point in the training set. Its boundary contorts in a manner that attempts to gain a few more points of accuracy. Two specific points, identified by black arrows, appear to drive much of the additional structure in the black curve. Since we know the actual boundary, it's easy to see that these contortions will not reproduce with new data. For example, it is implausible that a small island of vermilion circles exists in the mainstream of blue triangles. 

Another sign that a model probably has too much complexity is the additional component of the decision boundary at the bottom of @fig-two-class-overfit(d) (red arrow). Points inside this part of the boundary would be classified as triangles even though no triangles are near it. This is a sort of "blowback" of an overly complex model where choices the model makes in one part of the predictor space add complexity in a completely different part of the space. This blowback is often seen in areas where there are no data. It’s unnecessary and points to the idea that the model is trying to do too much. 

> This boundary demonstrates how added complexity can be a bad thing. A test set of data was simulated and is shown in @fig-two-class-new. The performance for this overly complex model is poor since the extra areas induced by the high cost value do not contain the right class of points. 

Reviewer: ```suggestion
This boundary demonstrates how added complexity can be a bad thing. A test set of data was simulated and is shown in @fig-two-class-new. The performance for this overly complex model is far from optimal since the extra areas induced by the high cost value do not contain the right class of points. 
```

#| warning: false
#| out-width: 35%
#| fig-width: 4
#| fig-height: 6
#| fig-cap: The high complexity fit from @fig-two-class-overfit(d) superimposed over a test set of data.

sim_new %>%
  ggplot(aes(A, B)) +
  geom_point(
    aes(col = class, pch = class),
    alpha = 3 / 4,
    cex = 2,
    show.legend = FALSE
  ) +
  coord_equal() +
  geom_contour(
    data = grid,
    aes(z = lp),
    breaks = 0,
    col = "black",
    linewidth = 2,
    alpha = 1 / 12
  ) +
  lims(x = c(-3, 3), y = c(-4, 4)) +
  labs(x = "Predictor A", y = "Predictor B")+
  geom_contour(
    data = high_cost,
    aes(z = value),
    breaks = 0,
    col = "black",
    linewidth = 1,
    alpha = 3 / 4
  )  +
  labs(x = "Predictor A", y = NULL) +
  scale_color_manual(values = sim_cols)
```

Our goal is to find tuning parameter values that are _just right_: complex enough to accurately represent the data but not complex enough to over-interpret it. 	
	
## The Ways Models Overfit {#sec-how-to-overfit}

This leads us to a formal definition of overfitting: 

::: {.important-box}

Overfitting is the situation where a model over-interprets trends and patterns in the training set that do not generalize to new data. 

> These irreproducible data trends result in models that **only do well on the training set**. 

Reviewer: ```suggestion
These irreproducible data trends result in models that **do very well on the training set** but greatly underperform on the validation set. 
```

Another example is overfitting the predictor set. The "low $N$, high $P$" problem [@Johnstone2009] occurs when data points are expensive to obtain, but each comes with abundant information. For example, in high-dimensional biology data, it is common to have dozens of data points and hundreds of thousands of predictors. Many models cannot be estimated with this imbalance of dimensions, and often, the first step is to filter out uninformative predictors. 

Suppose some statistical inferential method, such as a t-test, is used to determine if a predictor can differentiate between two classes. Small sample sizes make the chance of uninformative predictors falsely passing the filter large. If the same data are then used to build and evaluate the model, there is a significant probability that an ineffective model will appear to be very predictive [@Ambroise2002p1493;@fes]. Again, using a data set external to the model training and development process shows you when this occurs. 

> Finally, it is possible to overfit a model via sample filtering. For example, in some machine learning competitions, participants might sub-sample the training set to be as similar to the unlabelled test set as possible. This will most likely improve their test set performance, but it generally handicaps a model for other data sets. 

Reviewer: ```suggestion
Finally, it is possible to overfit a model via sample filtering. For example, in some machine learning competitions, participants might sub-sample the training set to be as similar to the unlabelled test set as possible. This will most likely improve their test set performance, but it generally handicaps a model when predicting a new yet-to-be-seen data set. 
```

## External Data to Measure Effectiveness {#sec-external-validation}

> By now, it should seem clear that the best way of avoiding overfitting is to quantify the model using a separate data set. How do we do that? We've said that most modeling projects create an initial partition of the data into training and testing sets. Can the testing set detect overfitting? 

Reviewer: ```suggestion
By now, it should seem clear that the best way of avoiding overfitting is to quantify the model performance using a separate data set. How do we do that? We've said that most modeling projects create an initial partition of the data into training and testing sets. Can the testing set detect overfitting? 
```

> We’ll again use the classification data from @fig-two-class-overfit to demonstrate, this time with a different model. A boosted tree, described in more detail later, is a collection of individual decision trees created sequentially. Each depends on the previous trees. For a new sample, each tree is predicted, and a weighted average is used to create the predicted class probabilities. We’ll use 100 trees to create our ensemble^[But, obviously, this is yet another tuning parameter.]. However, we don’t know how deep they should be. One method of controlling the number of splits in a tree is to control how much data should be in a partition to continue splitting. For example, in @fig-reg-tree, the left-most terminal node in the tree contained $n_{tr}$ = 971 data points after all splits are applied. If we just keep splitting with no constraint, at some point, the amount of data in the node is too small to continue. 

Reviewer: ```suggestion
We’ll again use the classification data from @fig-two-class-overfit to demonstrate, this time with a different model. A boosted tree, described in more detail later, is a collection of individual decision trees created sequentially. Each depends on the previous trees. For a new sample, each tree is used to predict the outcome, and a weighted average is used to create the predicted class probabilities. We’ll use 100 trees to create our ensemble^[But, obviously, this is yet another tuning parameter.]. However, we don’t know how deep the trees should be. One method of controlling the number of splits in a tree is to control how much data should be in a partition to continue splitting. For example, in @fig-reg-tree, the left-most terminal node in the tree contained $n_{tr}$ = 971 data points after all splits are applied. If we just keep splitting with no constraint, at some point, the amount of data in the node is too small to continue. 
```

```{r}
#| label: fig-complexity-resampling
#| echo: false
#| out-width: 60%
#| fig-width: 6
#| fig-height: 4.25
#| fig-cap: Model complexity versus model error with external validation via resampling.

brier_all %>% 
  ggplot(aes(min_n, brier, col = Method, pch = Method)) + 
  geom_point(cex = 1.2) + 
  geom_line() +
  labs(x = "Minimal Node Size", y = "Brier Score") +
  scale_color_manual(values = method_cols)
```	

We suggest using either of these two approaches (validation set or resampling) to measure performance during model development. Before going to the next chapter, let's give a preview of [Chapters -@sec-grid-search] and [-@sec-iterative-search].

## How Should We Optimize Complexity? {#sec-optimizing-complexity}

Now that we know how to appropriately estimate metrics for evaluating effectiveness, how do we choose tuning parameter values? Do we pick them at random? 

> Operationally, there are two main classes of search routines for optimizing model complexity via tuning parameters. The oldest is _grid search_. For each tuning parameter in the model, we define a reasonable set of specific candidate values to investigate. When there is more than one tuning parameter, there are various ways to create a multidimensional grid (these are discussed in @sec-grid-search). Each candidate in the grid is evaluated and the "best" combination of parameter values is chosen as the ones to use (if we think the model works well enough). 

Reviewer: ```suggestion
Operationally, there are two main classes of search routines for optimizing model complexity via tuning parameters. The oldest is _grid search_. For each tuning parameter in the model, we define a reasonable set of specific candidate values to investigate. When there is more than one tuning parameter, there are various ways to create a multidimensional grid (these are discussed in @sec-grid-search). Each candidate in the grid is evaluated and the "best" combination of parameter values is chosen as the one to use (if we think the model works well enough). 
```
